io/StacReader.cpp

```diff
@@ -0,0 +1,521 @@
+void StacReader::initializeItem(NL::json stacJson)
+{
+    if (prune(stacJson))
+        return;
+
+    if (m_args->schemaValidate)
+        schemaValidate(stacJson);
+
+    if (!stacJson["assets"].contains(m_args->assetName))
+        throw pdal_error("asset_name("+m_args->assetName+") doesn't match STAC object.");
+
+    std::string dataUrl = stacJson["assets"][m_args->assetName]["href"].get<std::string>();
+    std::string driver = m_factory.inferReaderDriver(dataUrl);
+
+    log()->get(LogLevel::Debug) << "Using driver " << driver <<
+        " for file " << dataUrl << std::endl;
+
+    Stage *reader = PluginManager<Stage>::createObject(driver);
+
+    if (!reader)
+        throwError("Unable to create reader for file '" + dataUrl + "'.");
+
+    Options readerOptions;
+    // add reader options defined in reader args to their respective readers
+    if (m_readerArgs.contains(driver)) {
+        NL::json args = m_readerArgs[driver].get<NL::json>();
+        for (auto& arg : args.items()) {
+            readerOptions.add(arg.key(), arg.value());
+        }
+    }
+
+    readerOptions.add("filename", dataUrl);
+    reader->setOptions(readerOptions);
+
+    if (m_readerList.size() > 0)
+        reader->setInput(*m_readerList.back());
+
+    if (reader)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_readerList.push_back(std::unique_ptr<Stage>(reader));
+    }
+}
+
+void StacReader::initializeCatalog(NL::json stacJson)
+{
+    if (m_args->schemaValidate)
+        schemaValidate(stacJson);
+    auto itemLinks = stacJson["links"];
+    for (auto link: itemLinks)
+    {
+        std::string linkType = link["rel"];
+        if (linkType != "item")
+            continue;
+        std::string itemUrl = link["href"];
+        //Create json from itemUrl
+        NL::json itemJson = NL::json::parse(m_arbiter->get(itemUrl));
+        initializeItem(itemJson);
+    }
+}
+
+// returns true if property matches
+bool matchProperty(std::string key, NL::json val, NL::json properties, NL::detail::value_t type)
+{
+    switch (type)
+    {
+        case NL::detail::value_t::string:
+        {
+            std::regex desired(val);
+            std::string value = properties[key].get<std::string>();
+            if (!std::regex_match(value, desired))
+                return false;
+            break;
+        }
+        case NL::detail::value_t::number_unsigned:
+        {
+            uint64_t value = properties[key].get<uint64_t>();
+            uint64_t desired = val.get<uint64_t>();
+            if (value != desired)
+                return false;
+            break;
+        }
+        case NL::detail::value_t::number_integer:
+        {
+            int value = properties[key].get<int>();
+            int desired = val.get<int>();
+            if (value != desired)
+                return false;
+            break;
+        }
+        case NL::detail::value_t::number_float:
+        {
+            int value = properties[key].get<int>();
```

Comment:
Is this bad copypasta? You're in the float case but fetching and asking for ints...

Reply:
Whoops, yes.